Enforcer rule for spring-cloud-gcp 1.1.2-RELEASE fails with missing dependency

While waiting for #769, I tried an old version of spring-cloud-gcp that does not depends on spring-core 5.2.0.M3. It fails as below:

```
[INFO] --- maven-enforcer-plugin:3.0.0-M2:enforce (enforce-linkage-checker) @ spring-cloud-gcp-core ---
[WARNING] The POM for org.apache.logging.log4j:log4j-api-java9:zip:2.11.2 is missing, no dependency information available
[WARNING] The POM for org.apache.logging.log4j:log4j-core-java9:zip:2.11.2 is missing, no dependency information available
[WARNING] The POM for xerces:xerces-impl:jar:2.6.2 is missing, no dependency information available
[WARNING] The POM for org.jruby.joni:joni:jar:2.1.1-SNAPSHOT is missing, no dependency information available
[ERROR] Could not find artifact org.apache.logging.log4j:log4j-api-java9:zip:2.11.2
[ERROR] Paths to the missing artifact:
 org.springframework.cloud:spring-cloud-gcp-core:jar:1.1.2.RELEASE
 > org.springframework:spring-core:jar:5.1.8.RELEASE (compile)
 > org.springframework:spring-jcl:jar:5.1.8.RELEASE (compile)
 > org.apache.logging.log4j:log4j-api:jar:2.11.2 (compile?)
 > org.apache.logging.log4j:log4j-api-java9:zip:2.11.2 (provided?)
```

Investigating the error.

After zip filtering (#811), it still has error for Netty:

```
[ERROR] Could not find artifact io.netty:netty-tcnative:jar:${os.detected.classifier}:2.0.25.Final
[ERROR] Paths to the missing artifact:
 org.springframework.cloud:spring-cloud-gcp-core:jar:1.1.2.RELEASE
 > org.springframework:spring-core:jar:5.1.8.RELEASE (compile)
 > io.projectreactor:reactor-core:jar:3.2.10.RELEASE (compile?)
 > io.micrometer:micrometer-core:jar:1.1.5 (compile?)
 > io.projectreactor.netty:reactor-netty:jar:0.8.9.RELEASE (compile?)
 > io.netty:netty-handler:jar:4.1.36.Final (compile?)
 > io.netty:netty-tcnative:jar:${os.detected.classifier}:2.0.25.Final (compile?
```

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java
@@ -20,6 +20,7 @@ import static com.google.common.collect.ImmutableList.toImmutableList;
 
 import com.google.common.base.CharMatcher;
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
 import java.util.ArrayDeque;
 import java.util.ArrayList;
 import java.util.HashMap;
@@ -57,22 +58,25 @@ public class DependencyGraphBuilder {
       CharMatcher.inRange('a', 'z').or(CharMatcher.inRange('0', '9'));
 
   static {
-    setDetectedOsSystemProperties();
+    detectOsProperties().forEach(System::setProperty);
   }
 
   // caching cuts time by about a factor of 4.
   private static final Map<String, DependencyNode> cacheWithProvidedScope = new HashMap<>();
   private static final Map<String, DependencyNode> cacheWithoutProvidedScope = new HashMap<>();
 
-  private static void setDetectedOsSystemProperties() {
+  public static ImmutableMap<String, String> detectOsProperties() {
     // System properties to select Netty dependencies through os-maven-plugin
     // Definition of the properties: https://github.com/trustin/os-maven-plugin
-
     String osDetectedName = osDetectedName();
-    System.setProperty("os.detected.name", osDetectedName);
     String osDetectedArch = osDetectedArch();
-    System.setProperty("os.detected.arch", osDetectedArch);
-    System.setProperty("os.detected.classifier", osDetectedName + "-" + osDetectedArch);
+    return ImmutableMap.of(
+        "os.detected.name",
+        osDetectedName,
+        "os.detected.arch",
+        osDetectedArch,
+        "os.detected.classifier",
+        osDetectedName + "-" + osDetectedArch);
   }
 
   private static String osDetectedName() {

--- a/enforcer-rules/pom.xml
+++ b/enforcer-rules/pom.xml
@@ -109,5 +109,10 @@
       <artifactId>mockito-core</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>com.google.truth</groupId>
+      <artifactId>truth</artifactId>
+      <scope>test</scope>
+    </dependency>
   </dependencies>
 </project>

--- a/enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java
+++ b/enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java
@@ -26,6 +26,7 @@ import com.google.cloud.tools.opensource.classpath.ClassReferenceGraph;
 import com.google.cloud.tools.opensource.classpath.LinkageChecker;
 import com.google.cloud.tools.opensource.classpath.SymbolProblem;
 import com.google.cloud.tools.opensource.dependencies.Artifacts;
+import com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilder;
 import com.google.cloud.tools.opensource.dependencies.FilteringZipDependencySelector;
 import com.google.cloud.tools.opensource.dependencies.NonTestDependencySelector;
 import com.google.common.annotations.VisibleForTesting;
@@ -217,6 +218,11 @@ public class LinkageCheckerRule extends AbstractNonCacheableEnforcerRule {
           helper.getComponent(ProjectDependenciesResolver.class);
       DefaultRepositorySystemSession fullDependencyResolutionSession =
           new DefaultRepositorySystemSession(session);
+
+      // For netty-handler referencing its dependencies with ${os.detected.classifier}
+      fullDependencyResolutionSession.setSystemProperties(
+          DependencyGraphBuilder.detectOsProperties());
+
       fullDependencyResolutionSession.setDependencySelector(
           new AndDependencySelector(
               new NonTestDependencySelector(),

--- a/enforcer-rules/src/test/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRuleTest.java
+++ b/enforcer-rules/src/test/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRuleTest.java
@@ -191,7 +191,7 @@ public class LinkageCheckerRuleTest {
         ArgumentCaptor.forClass(DependencyResolutionRequest.class);
     verify(mockProjectDependenciesResolver).resolve(argumentCaptor.capture());
     Truth.assertWithMessage(
-            "RepositorySystemSession should have variables defined in os-maven-plugin")
+            "RepositorySystemSession should have variables such as os.detected.classifier")
         .that(argumentCaptor.getValue().getRepositorySession().getSystemProperties())
         .containsAtLeastEntriesIn(DependencyGraphBuilder.detectOsProperties());
   }

--- a/enforcer-rules/src/test/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRuleTest.java
+++ b/enforcer-rules/src/test/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRuleTest.java
@@ -190,7 +190,9 @@ public class LinkageCheckerRuleTest {
     ArgumentCaptor<DependencyResolutionRequest> argumentCaptor =
         ArgumentCaptor.forClass(DependencyResolutionRequest.class);
     verify(mockProjectDependenciesResolver).resolve(argumentCaptor.capture());
-    Truth.assertThat(argumentCaptor.getValue().getRepositorySession().getSystemProperties())
+    Truth.assertWithMessage(
+            "RepositorySystemSession should have variables defined in os-maven-plugin")
+        .that(argumentCaptor.getValue().getRepositorySession().getSystemProperties())
         .containsAtLeastEntriesIn(DependencyGraphBuilder.detectOsProperties());
   }
 
@@ -435,5 +437,4 @@ public class LinkageCheckerRuleTest {
                   + "aopalliance:aopalliance:jar:1.0 (compile)");
     }
   }
-
 }

--- a/enforcer-rules/src/test/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRuleTest.java
+++ b/enforcer-rules/src/test/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRuleTest.java
@@ -25,9 +25,11 @@ import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
 import com.google.cloud.tools.dependencies.enforcer.LinkageCheckerRule.DependencySection;
+import com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilder;
 import com.google.cloud.tools.opensource.dependencies.RepositoryUtility;
 import com.google.common.collect.ImmutableList;
 import com.google.common.graph.Traverser;
+import com.google.common.truth.Truth;
 import java.net.URISyntaxException;
 import java.net.URLClassLoader;
 import java.nio.file.Paths;
@@ -64,6 +66,7 @@ import org.eclipse.aether.transfer.ArtifactNotFoundException;
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
+import org.mockito.ArgumentCaptor;
 import org.mockito.ArgumentMatchers;
 
 public class LinkageCheckerRuleTest {
@@ -174,6 +177,21 @@ public class LinkageCheckerRuleTest {
     // This should not raise an EnforcerRuleException
     rule.execute(mockRuleHelper);
     verify(mockLog).info("No error found");
+  }
+
+  @Test
+  public void testExecute_shouldPassGoodProject_sessionProperties()
+      throws EnforcerRuleException, RepositoryException, URISyntaxException,
+          DependencyResolutionException {
+    setupMockDependencyResolution("com.google.guava:guava:27.0.1-jre");
+
+    rule.execute(mockRuleHelper);
+
+    ArgumentCaptor<DependencyResolutionRequest> argumentCaptor =
+        ArgumentCaptor.forClass(DependencyResolutionRequest.class);
+    verify(mockProjectDependenciesResolver).resolve(argumentCaptor.capture());
+    Truth.assertThat(argumentCaptor.getValue().getRepositorySession().getSystemProperties())
+        .containsAtLeastEntriesIn(DependencyGraphBuilder.detectOsProperties());
   }
 
   @Test
@@ -417,4 +435,5 @@ public class LinkageCheckerRuleTest {
                   + "aopalliance:aopalliance:jar:1.0 (compile)");
     }
   }
+
 }

